### ISSUE TITLE
Training Time

### ISSUE THREAD

**Issue by rhalder2023** [ultralytics/ultralytics#19298]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, so i have trained and tuned the yolov8 model quite a lot of times before, but i am tuning it again now and somehow, now each epoch is taking around 8 minutes where as previously it used to take around 3 mins, on the same system specification. Does anyone know why this is happening, im on the 8.3.17 version of ultralytics.

**Comment by Y-T-G** [comment]
Try latest version

**Comment by rhalder2023** [comment]
> Try latest version

I just did like its 8.3.76 now and torch is on 2.1.2, if i update torch while its tuning does it have cause any problem or is ultralytics going to use a cached version of the previous torch till i rerun the code? I have never thought about this

**Comment by glenn-jocher** [comment]
@rhalder2023 **Answer:** Updating PyTorch mid-tuning won’t affect the current process (old version remains cached), but restart your code to use the new version. Always verify compatibility between Ultralytics and PyTorch versions to avoid issues. For more details, see our [YOLO Common Issues Guide](https://docs.ultralytics.com/guides/yolo-common-issues/).

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by rhalder2023** [comment]
> Can you provide the output after running this command in terminal: `yolo checks`?

![Image](https://github.com/user-attachments/assets/e43760c5-d8f3-4ec6-b450-e4c01076b060)

Output

**Comment by Y-T-G** [comment]
What's the training command used

**Comment by rhalder2023** [comment]
im tuning it, so

model.tune(data='/home/taiml3/Documents/Machine_Learning_Tools/yolov8/fused_model_data/data.yaml',name="Fused_Model_FirstTuning",            project="/home/taiml3/Documents/Machine_Learning_Tools/yolov8/Fused_Detection_Model", epochs=10, iterations=200, optimizer="Adam", 
           plots=False, save=False, val=True, batch=16)

its this

**Comment by rhalder2023** [comment]
> What's the training command used

btw can we like pause the tuning process?? or can i use the best hyperparamers file to start tuning from that point onwards?

**Comment by glenn-jocher** [comment]
You can't pause the tuning process, but you can start a new tuning process using the best hyperparameters from a previous run by specifying the `space` argument with the desired values. See [Tuner](https://docs.ultralytics.com/reference/engine/tuner/) for usage examples.

**Comment by rhalder2023** [comment]
> You can't pause the tuning process, but you can start a new tuning process using the best hyperparameters from a previous run by specifying the `space` argument with the desired values. See [Tuner](https://docs.ultralytics.com/reference/engine/tuner/) for usage examples.

So i did see that, but does it need like a range in the form of a tuple or can i just mention a starting value?

**Comment by glenn-jocher** [comment]
To specify fixed values (not ranges) when restarting tuning with previous best hyperparameters, provide each parameter as a tuple with identical min/max values in the `space` argument. For example: `space={"lr0": (0.001, 0.001)}` would fix `lr0` at 0.001. The Tuner documentation shows the required tuple format for parameter constraints.

**Comment by rhalder2023** [comment]
> To specify fixed values (not ranges) when restarting tuning with previous best hyperparameters, provide each parameter as a tuple with identical min/max values in the `space` argument. For example: `space={"lr0": (0.001, 0.001)}` would fix `lr0` at 0.001. The Tuner documentation shows the required tuple format for parameter constraints.

ah thanks

**Comment by glenn-jocher** [comment]
@rhalder2023 you're welcome! For clarity, each parameter in the `space` must be a tuple (even when fixed), as shown in the Tuner documentation. This ensures proper formatting for the hyperparameter constraints. Let us know if you encounter any issues!

**Comment by rhalder2023** [comment]
> [@rhalder2023](https://github.com/rhalder2023) you're welcome! For clarity, each parameter in the `space` must be a tuple (even when fixed), as shown in the Tuner documentation. This ensures proper formatting for the hyperparameter constraints. Let us know if you encounter any issues!

so if i had a lr tuple of (0.001,0.001) would that mean its not gonna update the lr during tuning making it fixed or will it still update the lr but without a fixed upper limit?

**Comment by rhalder2023** [comment]
> Yes, using `lr0: (0.001, 0.001)` in the `space` will fix the learning rate during tuning as the min/max bounds are identical. This parameter won't mutate. For details, see the [Hyperparameter Constraints section](https://docs.ultralytics.com/guides/hyperparameter-tuning/#custom-search-space-example) in our tuning guide.

Thank you for that. 

So umm i wanted to try multi gpu training/tuning, so i just passed the device parameter with [0,1] as the argument to the tune method call, and nvtop does show me both my gpus are being utilized but the time taken for each epoch to complete, is still the same, so i just wanted to know am i doing something wrong, or how does ultralytics utilize multiple gpus for training?

**Comment by rhalder2023** [comment]
> > Yes, using `lr0: (0.001, 0.001)` in the `space` will fix the learning rate during tuning as the min/max bounds are identical. This parameter won't mutate. For details, see the [Hyperparameter Constraints section](https://docs.ultralytics.com/guides/hyperparameter-tuning/#custom-search-space-example) in our tuning guide.
> 
> Thank you for that.
> 
> So umm i wanted to try multi gpu training/tuning, so i just passed the device parameter with [0,1] as the argument to the tune method call, and nvtop does show me both my gpus are being utilized but the time taken for each epoch to complete, is still the same, so i just wanted to know am i doing something wrong, or how does ultralytics utilize multiple gpus for training?

So the two gpus are different, one of them is the 4070 super ti and the other one is the 3060ti, is it because of the difference in compute capability of the two gpus, that caluclating the gradient takes longer in the 3060 and hence because its ddp and it needs to average the gradients its getting bottlenecked while the 4070 is waiting for the 3060?

**Comment by rhalder2023** [comment]
> > > Yes, using `lr0: (0.001, 0.001)` in the `space` will fix the learning rate during tuning as the min/max bounds are identical. This parameter won't mutate. For details, see the [Hyperparameter Constraints section](https://docs.ultralytics.com/guides/hyperparameter-tuning/#custom-search-space-example) in our tuning guide.
> > 
> > 
> > Thank you for that.
> > So umm i wanted to try multi gpu training/tuning, so i just passed the device parameter with [0,1] as the argument to the tune method call, and nvtop does show me both my gpus are being utilized but the time taken for each epoch to complete, is still the same, so i just wanted to know am i doing something wrong, or how does ultralytics utilize multiple gpus for training?
> 
> So the two gpus are different, one of them is the 4070 super ti and the other one is the 3060ti, is it because of the difference in compute capability of the two gpus, that caluclating the gradient takes longer in the 3060 and hence because its ddp and it needs to average the gradients its getting bottlenecked while the 4070 is waiting for the 3060?

I ran some more tests, it takes around 13 mins per epoch on the 3060, around 6 mins 50 seconds on the 4070 and around 7 mins when both gpus are used, it might be the communication overhead, or the bandwidth between the gpus, so i was thinking of giving device weights like use 70% of the data on the 4070 and 30% of the data on the 3060, can this be done through the ultralytics api or do i have to use the torch method for it.

**Comment by glenn-jocher** [comment]
**Answer:** YOLO11 uses PyTorch DDP for multi-GPU training, which synchronizes gradients across devices. Mixing GPUs with different compute capabilities (like 4070 and 3060Ti) can create bottlenecks as all devices wait for the slowest one. While Ultralytics doesn't natively support weighted device allocation, we recommend using GPUs with similar speeds. For custom implementations, you'd need to modify the PyTorch dataloader directly. See our [Multi-GPU Training Guide](https://docs.ultralytics.com/guides/yolo-common-issues/#accelerating-training-with-multiple-gpus) for best practices.

**Comment by rhalder2023** [comment]
> **Answer:** YOLO11 uses PyTorch DDP for multi-GPU training, which synchronizes gradients across devices. Mixing GPUs with different compute capabilities (like 4070 and 3060Ti) can create bottlenecks as all devices wait for the slowest one. While Ultralytics doesn't natively support weighted device allocation, we recommend using GPUs with similar speeds. For custom implementations, you'd need to modify the PyTorch dataloader directly. See our [Multi-GPU Training Guide](https://docs.ultralytics.com/guides/yolo-common-issues/#accelerating-training-with-multiple-gpus) for best practices.

And its the same for v8 right cause thats what im using

**Comment by glenn-jocher** [comment]
**Answer:** Yes, the same PyTorch DDP principles apply to YOLOv8. For optimal multi-GPU performance with any YOLO version, we recommend matching GPU models to avoid bottlenecks. The [Multi-GPU Training Guide](https://docs.ultralytics.com/guides/yolo-common-issues/#accelerating-training-with-multiple-gpus) applies universally across all supported YOLO versions.

**Comment by rhalder2023** [comment]
> You're welcome! For optimal multi-GPU performance with YOLOv8, we strongly recommend using GPUs with matching compute capabilities as outlined in our Multi-GPU Training Guide. Let us know if you encounter any other issues! 🚀

so yeah, i have an input image of 360,640 but my predicted masks are of shape 384,640  

why is this happening?

**Comment by glenn-jocher** [comment]
The mask output size (384x640) aligns with YOLOv8's 32-pixel stride requirement for input dimensions. Since 360 isn't divisible by 32, the model automatically pads inputs to the nearest multiple (384). This ensures proper feature map alignment during segmentation. You can maintain original dimensions by using input sizes divisible by 32, like 352x640 (352 = 32*11). For details see the [Segmentation Task Guide](https://docs.ultralytics.com/tasks/segment/).

**Comment by rhalder2023** [comment]
> The mask output size (384x640) aligns with YOLOv8's 32-pixel stride requirement for input dimensions. Since 360 isn't divisible by 32, the model automatically pads inputs to the nearest multiple (384). This ensures proper feature map alignment during segmentation. You can maintain original dimensions by using input sizes divisible by 32, like 352x640 (352 = 32*11). For details see the [Segmentation Task Guide](https://docs.ultralytics.com/tasks/segment/).

Yeah i realized that last night, Thank You, so these pixels are just padded at the end of the array and i can just drop them to get the original shape?

**Comment by glenn-jocher** [comment]
@rhalder2023 yes, you can remove the padded pixels to recover the original dimensions. The [scale_masks function](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.scale_masks) in our ops module automatically handles this padding reversal when processing outputs. For optimal results without padding, use input resolutions divisible by 32 as recommended in our [image size check function](https://docs.ultralytics.com/reference/utils/checks/#ultralytics.utils.checks.check_imgsz).

**Comment by rhalder2023** [comment]
> [@rhalder2023](https://github.com/rhalder2023) yes, you can remove the padded pixels to recover the original dimensions. The [scale_masks function](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.scale_masks) in our ops module automatically handles this padding reversal when processing outputs. For optimal results without padding, use input resolutions divisible by 32 as recommended in our [image size check function](https://docs.ultralytics.com/reference/utils/checks/#ultralytics.utils.checks.check_imgsz).

So i checked the output image, its getting padded 12 on the front and 12 at the back, is that how it works?

**Comment by glenn-jocher** [comment]
Yes, the padding is split equally on both ends when possible (as seen in the `scale_masks` function's `pad[0] /= 2` and `pad[1] /= 2` logic). For your 360→384 padding case with 24px total, this results in 12px added to both top/bottom or left/right. The [scale_masks source code](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.scale_masks) shows how padding is calculated and applied.

**Comment by rhalder2023** [comment]
> Yes, the padding is split equally on both ends when possible (as seen in the `scale_masks` function's `pad[0] /= 2` and `pad[1] /= 2` logic). For your 360→384 padding case with 24px total, this results in 12px added to both top/bottom or left/right. The [scale_masks source code](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.scale_masks) shows how padding is calculated and applied.

Thank you for this.

So i wanted to try v12, and regarding the flash attention, is there like some sort of way to know that flash attnetion has been compiled succesfully through ultralytics, like u guys have the ultralytics.checks() method, or should i use torch to verify if its been compiled successfully or not?

**Comment by glenn-jocher** [comment]
@rhalder2023 for verifying FlashAttention in YOLO12, check if `torch.backends.cuda.flash_sdp_enabled()` returns `True` after model initialization. The Ultralytics YOLO12 implementation will log "FlashAttention enabled" during startup if successfully compiled. For details, see the [YOLO12 Requirements section](https://docs.ultralytics.com/models/yolo12/#requirements).

**Comment by rhalder2023** [comment]
> [@rhalder2023](https://github.com/rhalder2023) for verifying FlashAttention in YOLO12, check if `torch.backends.cuda.flash_sdp_enabled()` returns `True` after model initialization. The Ultralytics YOLO12 implementation will log "FlashAttention enabled" during startup if successfully compiled. For details, see the [YOLO12 Requirements section](https://docs.ultralytics.com/models/yolo12/#requirements).

so do i just download the package from dao-ailabs? or is there like a preferred method of installation?

**Comment by rhalder2023** [comment]
> [@rhalder2023](https://github.com/rhalder2023) for verifying FlashAttention in YOLO12, check if `torch.backends.cuda.flash_sdp_enabled()` returns `True` after model initialization. The Ultralytics YOLO12 implementation will log "FlashAttention enabled" during startup if successfully compiled. For details, see the [YOLO12 Requirements section](https://docs.ultralytics.com/models/yolo12/#requirements).

Do u guys not a have a yolo12-seg trained on the coco dataset, do i gotta train using the yaml file???

**Comment by glenn-jocher** [comment]
For YOLO12 segmentation models, you'll need to train from the provided YAML configuration. You can initialize training using a pretrained detection model like `model = YOLO('yolo12n.pt').train(data='coco-seg.yaml')`. For details, see the Segmentation Training Guide and Model Training documentation.
